### PR TITLE
Multi-host tablespace tests

### PR DESCRIPTION
**Pull request by kalensk**
This is a followup to PR https://github.com/greenplum-db/gpdb/pull/7484 to add multi-host (`concourse_cluster`) table space tests.

Note: A lot of these `concourse_cluster` tests are duplicates or very similar to their `demo_cluster` counterparts. This is quickly starting to be a pain point that can be solved by writing a test once, and having it seamlessly run on both a `demo_cluster` and `concourse_cluster`.

We are currently skipping gprecoverseg tablespace tests as tablespaces do not work with pg_rewind yet. PR #7473 helps address that with another followup PR needed. A followup chore to remove the skip tag for the gprecoverseg tests has been added.

A full test pipeline is being run. And a 6X backport is needed.

### CODE REVIEW

File: concourse/pipelines/templates/gpdb-tpl.yml [hunk 1]
```diff
@@ -15,7 +15,7 @@
                       {'name': 'replication_slots',
                        'use_concourse_cluster': false},
                       {'name': 'gpactivatestandby',
-                       'use_concourse_cluster': false},
+                       'use_concourse_cluster': true},
```
Comment (dkrieger-pvtl) [inline concourse/pipelines/templates/gpdb-tpl.yml]: do we want to add `'additional_ccp_vars': 'number_of_nodes: 4'` here?  The default of 2 for this gives us only one segment, and perhaps there's a different here.

File: gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature [hunk 2]
```diff
@@ -103,9 +103,26 @@ Feature: Tests for gpaddmirrors
         And gpaddmirrors adds mirrors with temporary data dir
         And an FTS probe is triggered
         And the segments are synchronized
-        When user kills all primary processes with SIGKILL
+        When user kills all primary processes with SIGTERM
```
Comment (jchampio) [inline gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature]: nit: this is identical to `user kills all primary processes`

File: gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature [hunk 3]
```diff
@@ -103,9 +103,26 @@ Feature: Tests for gpaddmirrors
...
+         When gpaddmirrors adds mirrors
+         Then verify the database has mirrors
+
+         When an FTS probe is triggered
+          And the segments are synchronized
+         Then the tablespace is valid
+
+         When user kills all primary processes
+          And user can start transactions
+         Then the tablespace is valid
```
Comment (jchampio) [inline gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature]: You'll want to set up your editor to ensure newlines exist at the end of files.

File: gpMgmt/test/behave/mgmt_utils/steps/tablespace_mgmt_utils.py [hunk 4]
```diff
@@ -15,6 +16,10 @@ def __init__(self, name):
         self.table_counter = 0
         self.initial_data = None
 
+        gparray = GpArray.initFromCatalog(dbconn.DbURL())
+        for host in gparray.getHostList():
+            run_cmd('ssh %s mkdir -p %s' % (host, self.path))
```
Comment (jchampio) [inline gpMgmt/test/behave/mgmt_utils/steps/tablespace_mgmt_utils.py]: `self.path` should be `pipes.quote(self.path)` to handle any funny business. Technically `host` needs to be quoted too, but since this is test code I'm not worried about that one.

It might be nice to point out in a comment that this directory is not automatically cleaned up for any hosts except the current one.

File: gpMgmt/test/behave/mgmt_utils/gpactivatestandby.feature [hunk 5]
```diff
@@ -74,3 +74,21 @@ Feature: gpactivatestandby
...
+########################### @concourse_cluster tests ###########################
+# The @concourse_cluster tag denotes the scenario that requires a remote cluster
+
+    @concourse_cluster
+    Scenario: tablespaces work on a multi-host environment
+        Given the database is running
+          And the standby is not initialized
+          And a tablespace is created with data
+         When the user runs gpinitstandby with options " "
+         Then gpinitstandby should return a return code of 0
```
Comment (dkrieger-pvtl) [inline gpMgmt/test/behave/mgmt_utils/gpactivatestandby.feature]: Shouldn't the `When` and `Then` above be `And`?  We're testing `gpactivatestandby` here, and `gpinitstandby` is just a precondition.

File: gpMgmt/test/behave/mgmt_utils/gprecoverseg.feature [hunk 6]
```diff
@@ -227,3 +227,41 @@ Feature: gprecoverseg tests
         And the segments are synchronized
         # validate the the new segment has the correct setting by getting admin connection to that segment
         Then the saved primary segment reports the same value for sql "show data_checksums" db "template1" as was saved
+
+    @skip  # tablespaces are being reworked and currently do not work with pg_rewind
```
Comment (jchampio) [inline gpMgmt/test/behave/mgmt_utils/gprecoverseg.feature]: Is there a chore to un-skip these?

File: gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature (hunk 3 above)
Comment (dkrieger-pvtl) [inline gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature]: nit: add a newline at the end of this file.

File: gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature (hunk 2 above)
Comment (dkrieger-pvtl) [inline gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature]: Why did you change this to `SIGTERM`?  This can lead to flakes, as if the primaries don't drop in time(30 seconds serially for each of the three primaries), this test will fail with an exception.

File: gpMgmt/test/behave/mgmt_utils/gprecoverseg.feature (hunk 6 above)
Comment (kalensk) [reply gpMgmt/test/behave/mgmt_utils/gprecoverseg.feature]: Yes, "Remove skip tag from gprecoverseg tablespace tests in master and 6X_STABLE" (#165426224)

File: gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature (hunk 2 above)
Comment (marksliva) [reply gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature]: We made this change to be as small as possible to get tablespaces working without having to refactor this unrelated step. If this starts to flake then we can address it then. Our longer term goal should be to replace kill with gpstop for graceful shutdowns. If need be, we can augment gpstop to target specific segments instead of the whole cluster.

File: gpMgmt/test/behave/mgmt_utils/steps/tablespace_mgmt_utils.py (hunk 4 above)
Comment (marksliva) [reply gpMgmt/test/behave/mgmt_utils/steps/tablespace_mgmt_utils.py]: Nice catch.

File: gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature (hunk 3 above)
Comment (marksliva) [reply gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature]: Is this a Python and/or behave file convention? Just curious..

File: gpMgmt/test/behave/mgmt_utils/gpactivatestandby.feature (hunk 5 above)
Comment (marksliva) [reply gpMgmt/test/behave/mgmt_utils/gpactivatestandby.feature]: Did you have something like this in mind?
```
        Given the database is running
          And the standby is not initialized
          And a tablespace is created with data
         When the user runs gpinitstandby with options " "
         Then gpinitstandby should return a return code of 0
          And verify the standby master entries in catalog

         When the master goes down
         Then the user runs gpactivatestandby with options " "
          And gpactivatestandby should return a return code of 0
          And verify the standby master is now acting as master
          And the tablespace is valid on the standby master
          And clean up and revert back to original master
```

File: concourse/pipelines/templates/gpdb-tpl.yml (hunk 1 above)
Comment (kalensk) [reply concourse/pipelines/templates/gpdb-tpl.yml]: Yes, `additional_ccp_vars` defaults to creating an `mdw` and `sdw1`. I would like to leave this to the default until we write a test that requires more nodes. That is, let the code decide.

File: gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature (hunk 3 above)
Comment (marksliva) [reply gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature]: We added a newline to the EOF

File: gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature (hunk 2 above)
Comment (marksliva) [reply gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature]: We have updated it to use `user kills all primary processes`

File: gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature (hunk 3 above)
Comment (marksliva) [reply gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature]: Addressed below

File: gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature (hunk 2 above)
Comment (kalensk) [reply gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature]: To add more details (as described in our commit message), shutting down the primaries using SIGKILL keeps the PID files around, whereas SIGTERM more gracefully stops the primaries and removes the PID files. When PID files hang around, it prevents future scenarios from being able to create a cluster (since it thinks a postgres process is already running).

Zombie pid files from SIGKILL was preventing our tablespace test from successfully creating a cluster, since it thought a postgres process was already running. This was the least invasive change we thought of to proceed with writing our test.

File: concourse/pipelines/templates/gpdb-tpl.yml (hunk 1 above)
Comment (dkrieger-pvtl) [reply concourse/pipelines/templates/gpdb-tpl.yml]: I was thinking more that, somehow, having two segment hosts(3 nodes total) might catch more issues with table spaces than one.  We are testing possible gpactivatestandby interactions with tablespaces... I am not sure what you mean by `let the code decide`...we use the current infrastructure to decide the number of notes.

File: gpMgmt/test/behave/mgmt_utils/gpactivatestandby.feature (hunk 5 above)
Comment (dkrieger-pvtl) [reply gpMgmt/test/behave/mgmt_utils/gpactivatestandby.feature]: I was thinking of this.  See the Gherkin discussion for Cucumber for background: https://cucumber.io/docs/gherkin/reference/.  Both above and below seem like valid interpretations of Gherkin, but to me, the preconditions are setting up state for what we are testing, which here is `gpactivatestandby`.  With the below, it's clear what exactly we are testing.

```gherkin
        Given the database is running
          And the standby is not initialized
          And a tablespace is created with data
          And the user runs gpinitstandby with options " "
          And gpinitstandby should return a return code of 0
          And verify the standby master entries in catalog
          And the master goes down
        When the user runs gpactivatestandby with options " "
         Then gpactivatestandby should return a return code of 0
          And verify the standby master is now acting as master
          And the tablespace is valid on the standby master
          And clean up and revert back to original master
```

File: gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature (hunk 2 above)
Comment (dkrieger-pvtl) [reply gpMgmt/test/behave/mgmt_utils/gpaddmirrors.feature]: Ok, sorry, I just reviewed the entire PR and I didn't real the individual commit comments.  For future reference, just another reason to use `gpstop` which appears to clean up properly even after a `SIGKILL`...but we might see some flakes here.  As a cleanup task, we should add a chore to cleanup our techniques for stopping our cluster in our Behave tests as we are open to flakes.  One easy way to test(once, or with the fault injector) might be to add a 60 second sleep in the shutdown path of the postmaster...

File: gpMgmt/test/behave/mgmt_utils/gpactivatestandby.feature (hunk 5 above)
Comment (dkrieger-pvtl) [reply gpMgmt/test/behave/mgmt_utils/gpactivatestandby.feature]: But this also can be deferred via a new chore task, as it's a wider discussion for all of our Behave tests.

Comment (kalensk) [reply gpMgmt/test/behave/mgmt_utils/gpactivatestandby.feature]: It seems like there is very little consistency or cohesion amongst when and how to use Give, When, Then in our tests. Perhaps we need a meeting and doc that we can all follow to start to give some consistency.

I was thinking something like the following definitions taken from [here](http://docs.behat.org/en/v2.5/guides/1.gherkin.html#steps):
- `Given` puts the system in a known state before the user (or external system) starts interacting with it
- `When` describe the key action the ***user*** performs 
- `Then` observe outcomes from when the user performs the action. The observations should be related to the business value/benefit in your feature description.
- Try to have one set of Given, When, Then per scenario, but it is "ok" to have multiple if needed separated by a blank line.
- Try to avoid dangling When, or Then's - that is try to have all three Given, When, and Then together. For example,
```
Given
And
When
Then

Given
When
Then
```

Rather than,

```
Given
And
When
Then

When
Then
And
```

Comment (dkrieger-pvtl) [reply gpMgmt/test/behave/mgmt_utils/gpactivatestandby.feature]: Yes, as a separate chore, we should decide on a convention/semantics of the behavioral testing words we use.
